tests/test_xlora.py

```diff
@@ -381,3 +382,126 @@ def test_xlora_loading_valid(self):
+
+                                    capture_info = {
+                                        "layer": layer_idx,
+                                        "projection": name,
+                                        "result_shape": result.shape if hasattr(result, "shape") else "unknown",
+                                        "timestamp": len(captured_data),
+                                        "normalized_scalings": normalized_scalings,
+                                    }
+                                    captured_data.append(capture_info)
+                                    return result
+                                else:
+                                    return orig_forward(*args, **kwargs)
+
+                            return wrapped_forward
+
+                        wrapper = create_wrapped_forward(original_forward, layer_idx, proj_name, xlora_wrapper)
+
+                        mp = pytest.MonkeyPatch()
+                        mp.setattr(proj_layer, "forward", wrapper)
+                        monkeypatches.append(mp)
+                        wrapped_count += 1
+
+            return wrapped_count
+
+        total_wrapped = wrap_decoder_layers()
+        assert total_wrapped > 0, "No X-LoRA layers were wrapped for testing."
+
+        try:
+            model.enable_scalings_logging()
+            inputs = tokenizer.encode("Test per token normalization", add_special_tokens=False, return_tensors="pt")
+            outputs = model.generate(
+                input_ids=inputs.to(self.torch_device),
+                max_new_tokens=1,
+            )
+
+            assert len(captured_data) > 0, "No scaling data was captured during forward pass."
+
+            for data in captured_data:
+                normalized_scalings = data.get("normalized_scalings")
+                if normalized_scalings is None:
+                    assert normalized_scalings is not None, (
+                        f"Missing normalized_scalings in layer {data['layer']} {data['projection']}"
+                    )
+                    continue
+
+                if hasattr(normalized_scalings, "cpu"):
+                    scalings_np = normalized_scalings.cpu().detach().numpy()
+                else:
+                    scalings_np = normalized_scalings
+
+                # expected shape: (batch, seq_len, num_loras) or (batch, seq_len, top_k)
+                assert scalings_np.ndim == 3, (
+                    f"Unexpected scalings shape {scalings_np.shape} in layer {data['layer']} {data['projection']}"
+                )
+
+                batch_size, seq_len, num_experts = scalings_np.shape
+                for b in range(batch_size):
+                    for t in range(seq_len):
+                        weights = scalings_np[b, t, :]
+                        weight_sum = weights.sum()
+                        assert np.isclose(weight_sum, 1.0, atol=1e-5), (
```

Comment:
I think this is the essential part of the test. I would focus on this assert, no need to check the other stuff and also no need to report the layer, batch, and token in detail.